# Problema del agricultor (granjero)

Considere un granjero que tiene $500$ hectáreas de tierra para cultivar grano, maíz y betabel. En invierno quiere decidir qué tanta tierra debe designar para cada cultivo con el objetivo de obtener la máxima ganancia el siguiente otoño. 
Así mismo el granjero necesita $200$ toneladas de trigo y $240$ toneladas de maíz para alimentar su ganado. Esta cantidad puede ser obtenida por cultivo o la puede comprar externamente. Si se produce más cantidad de la necesaria, la parte excedente se puede vender. Los precios de venta son de $\$170$ y $\$150$ por tonelada de trigo y maíz respectivamente. Los precios de compra son un $40\%$ menos que el precio de venta.


El betabel se vende a $\$36$ por tonelada. Sin embargo, hay una cuota para esta producción: Cualquier cantidad en exceso de la cuota puede ser solamente vendida en $\$10$ por tonelada. La cuota este año es de $6000$ toneladas. 


Basándose en experiencias pasadas, el granjero sabe que la producción promedia de sus cultivos es de $2.5$, $3$ y $10$ toneladas de trigo, maíz y betabel respectivamente. Así mismo, los costos de plantación son $\$150$, $\$230$ y $\$260$ por hectárea respectivamente. 

De manera resumida la información se puede ver en la Tabla 1. 

.


<center>**Tabla 1: Características del problema**</center>

 | Trigo | Maíz | Betabel
:-------| --
Producción (Toneladas/Hectárea) | 2.5 | 3  | 20
Costo ($\$$/Hectárea)| 150 | 230 |260 
Precio de Venta ($\$$/Tonelada) | 170 | 150 |  (cuota) 36 ó 10
Costo ($\$$/Tonelada)| 238|210|
Requerimiento mínimo (Tonelada)| 200|240| 
Tierra disponible: 500 hectáreas|

Con esta información se puede implementar la programación lineal para poder obtener resultados satisfactorios. 
Desde una perspectiva matemática, se le puede asignar las variables $x_1$, $x_2$ $x_3$ a la cantidad de tierra otorgada para cada cultivo de trigo, maíz y betabel respectivamente. Las variables $w_1$ y $w_2$ se le asignarán a las toneladas de trigo y maíz vendido; $y_1$ y $y_2$ serán las toneladas de trigo y maíz comprado. $w_3$ será la cantidad de toneladas de betabel vendidas a un precio favorable y $w_4$ la cantidad de toneladas de betabel vendidas aun precio menor. 

Básicamente:

* $x_i:$Número de hectáreas asignadas

* $w_i:$Toneladas vendidas

* $y_i:$Toneladas compradas

En este caso, se quiere minimizar el gasto neto, (lo que sería lo mismo a decir maximizar la ganacia neta), todo esto bajo una serie de restricciones.

La formulación matemática del problema se puede expresar de la siguiente manera:

$minimizar$ $150 x_1 + 230 x_2 + 260 x_3 + 238 y_1 + 210 y_2 - 170 w_1 - 150 w_2- 36 w_3 - 10 w_4 $

$tal$ $que$

$x_1 + x_2 + x_3 \leq 500$

$2.4 x_1 + y_1 - w_1 \geq 200$

$3 x_2 + y_2 - w_2 \geq 240$

$w_3 + w_4 \leq 20 x_3$

$w_3 \leq 6000$

$x_1, x_2, x_3,y_1, y_2, w_1, w_2, w_3, w_4 \geq 0$

In [1]:
#Programa: Problema del agricultor (granjero)

#Usar librerias
from pulp import *

#Se crea la variable de problema:
prob=LpProblem("Problema del agricultor (granjero)", LpMinimize)

#Se crean cada una de las variables que se requieren determinar
#Aquí se definen los límites de las variables:
x1=LpVariable("x1",lowBound=0)
x2=LpVariable("x2",lowBound=0)
x3=LpVariable("x3",lowBound=0)
y1=LpVariable("y1",lowBound=0)
y2=LpVariable("y2",lowBound=0)
w1=LpVariable("w1",lowBound=0)
w2=LpVariable("w2",lowBound=0)
w3=LpVariable("w3",lowBound=0)
w4=LpVariable("w4",lowBound=0)

#Se crea la funcion objetivo:
funcion_objetivo= 150*x1 + 230*x2 + 260*x3 + 238*y1 + 210*y2 - 170*w1 - 150*w2- 36*w3 - 10*w4 

#La función objetivo se agrega al problema
prob += funcion_objetivo

#Se crean las restricciones (en este caso solo se tienen 5 restricciones):
restriccion1= x1 + x2 + x3 <= 500
restriccion2= 2.5*x1 + y1 - w1 >= 200
restriccion3= 3*x2 + y2 - w2 >= 240
restriccion4= w3 + w4 - 20*x3 <= 0
restriccion5= w3 <= 6000

#Las restricciones se añaden
prob += restriccion1
prob += restriccion2
prob += restriccion3
prob += restriccion4
prob += restriccion5

#Se soluciona el problema:
prob.solve()

#Resultados impresos:
print ("Status:", LpStatus[prob.status])
for v in prob.variables():
    print(v.name, "=", v.varValue)
print("Valor Óptimo: $", value(prob.objective))

Status: Optimal
w1 = 100.0
w2 = 0.0
w3 = 6000.0
w4 = 0.0
x1 = 120.0
x2 = 80.0
x3 = 300.0
y1 = 0.0
y2 = 0.0
Valor Óptimo: $ -118600.0


En este caso se asume que la cosecha tiene un 100% de probabilidad de darse, no obsante esta puede variar un poco, por lo que se propone un escenario de casos.

Entonces se puede asumir que la producción de cada cultivo está correlacionada con el clima. Se resume que se tienen tres posibles condiciones: buena, mala y promedio. La condición buena es tener un 20% más que la producción de cada cultivo, y la condición mala es tener un 20% menos que la producción de cada cultivo.

Debido a esto, se obtienen tres distintos casos individuales:

Para el caso "bueno" la formulación matemática se puede observar de la siguiente manera:

$minimizar$ $150 x_1 + 230 x_2 + 260 x_3 + 238 y_1 + 210 y_2 - 170 w_1 - 150 w_2- 36 w_3 - 10 w_4 $

$tal$ $que$

$x_1 + x_2 + x_3 \leq 500$

$3 x_1 + y_1 - w_1 \geq 200$

$3.6 x_2 + y_2 - w_2 \geq 240$

$w_3 + w_4 \leq 24 x_3$

$w_3 \leq 6000$

$x_1, x_2, x_3,y_1, y_2, w_1, w_2, w_3, w_4 \geq 0$


In [2]:
#Escenario Bueno:
#Programa: Problema del agricultor (granjero)


#Usar librerias
from pulp import *

#Se crea la variable de problema:
prob=LpProblem("Problema del agricultor (granjero)", LpMinimize)

#Se crean cada una de las variables que se requieren determinar
#Aquí se definen los límites de las variables:
x1=LpVariable("x1",lowBound=0)
x2=LpVariable("x2",lowBound=0)
x3=LpVariable("x3",lowBound=0)
y1=LpVariable("y1",lowBound=0)
y2=LpVariable("y2",lowBound=0)
w1=LpVariable("w1",lowBound=0)
w2=LpVariable("w2",lowBound=0)
w3=LpVariable("w3",lowBound=0)
w4=LpVariable("w4",lowBound=0)

#Se crea la funcion objetivo:
funcion_objetivo= 150*x1 + 230*x2 + 260*x3 + 238*y1 + 210*y2 - 170*w1 - 150*w2- 36*w3 - 10*w4 

#La función objetivo se agrega al problema
prob += funcion_objetivo

#Se crean las restricciones (notese que solo se modifica la restricción 2, 3 y, 4):
restriccion1= x1 + x2 + x3 <= 500
restriccion2= 3*x1 + y1 - w1 >= 200
restriccion3= 3.6*x2 + y2 - w2 >= 240
restriccion4= w3 + w4 - 24*x3 <= 0
restriccion5= w3 <= 6000

#Las restricciones se añaden 
prob += restriccion1
prob += restriccion2
prob += restriccion3
prob += restriccion4
prob += restriccion5

#Se soluciona el problema:
prob.solve()

#Resultados impresos:
print ("Status:", LpStatus[prob.status])
for v in prob.variables():
    print(v.name, "=", v.varValue)
print("Valor Óptimo: $", value(prob.objective))

Status: Optimal
w1 = 350.0
w2 = 0.0
w3 = 6000.0
w4 = 0.0
x1 = 183.33333
x2 = 66.666667
x3 = 250.0
y1 = 0.0
y2 = 0.0
Valor Óptimo: $ -167666.66709


Para el caso "malo" la formulación matemática se puede observar de la siguiente manera:

$minimizar$ $150 x_1 + 230 x_2 + 260 x_3 + 238 y_1 + 210 y_2 - 170 w_1 - 150 w_2- 36 w_3 - 10 w_4 $

$tal$ $que$

$x_1 + x_2 + x_3 \leq 500$

$2 x_1 + y_1 - w_1 \geq 200$

$2.4 x_2 + y_2 - w_2 \geq 240$

$w_3 + w_4 \leq 16 x_3$

$w_3 \leq 6000$

$x_1, x_2, x_3,y_1, y_2, w_1, w_2, w_3, w_4 \geq 0$

Se debe notar que lo único que varía es la cantidad de producción de toneladas por hectárea anual. Es decir, en promedio se esperaría que fuesen $2.5$ de trigo, $3$ de maíz y $20$ de betabel, pero en realidad, estos podrían ser datos variables en el tiempo, es decir que tienen cierta probabilidad de disminuir o aumentar. En la siguiente tabla se muestran los respectivos cambios. 

Producción (Toneladas/Hectárea) | Trigo | Maíz | Betabel
:-------| --
Promedio| 2.5 | 3  | 20
Bueno| 3 | 3.6  | 24
Malo| 2 | 2.4  | 16


In [3]:
#Escenario Malo:
#Programa: Problema del agricultor (granjero)


#Usar librerias
from pulp import *

#Se crea la variable de problema:
prob=LpProblem("Problema del agricultor (granjero)", LpMinimize)

#Se crean cada una de las variables que se requieren determinar
#Aquí se definen los límites de las variables:
x1=LpVariable("x1",lowBound=0)
x2=LpVariable("x2",lowBound=0)
x3=LpVariable("x3",lowBound=0)
y1=LpVariable("y1",lowBound=0)
y2=LpVariable("y2",lowBound=0)
w1=LpVariable("w1",lowBound=0)
w2=LpVariable("w2",lowBound=0)
w3=LpVariable("w3",lowBound=0)
w4=LpVariable("w4",lowBound=0)

#Se crea la funcion objetivo:
funcion_objetivo= 150*x1 + 230*x2 + 260*x3 + 238*y1 + 210*y2 - 170*w1 - 150*w2- 36*w3 - 10*w4 

#La función objetivo se agrega al problema
prob += funcion_objetivo

#Se crean las restricciones (notese que solo se modifica la restricción 2, 3 y, 4):
restriccion1= x1 + x2 + x3 <= 500
restriccion2= 2*x1 + y1 - w1 >= 200
restriccion3= 2.4*x2 + y2 - w2 >= 240
restriccion4= w3 + w4 - 16*x3 <= 0
restriccion5= w3 <= 6000

#Las restricciones se añaden
prob += restriccion1
prob += restriccion2
prob += restriccion3
prob += restriccion4
prob += restriccion5

#Se soluciona el problema:
prob.solve()

#Resultados impresos:
print ("Status:", LpStatus[prob.status])
for v in prob.variables():
    print(v.name, "=", v.varValue)
print("Valor Óptimo: $", value(prob.objective))

Status: Optimal
w1 = 0.0
w2 = 0.0
w3 = 6000.0
w4 = 0.0
x1 = 100.0
x2 = 25.0
x3 = 375.0
y1 = 0.0
y2 = 180.0
Valor Óptimo: $ -59950.0


Debido a esto, se concluye que saber la condición climática antes de actuar es una acción necesaria para poder tomar una decisión ya que esta repercute en los resultados. No obstante esta información no puede ser predecida en anticipación, por lo que se debe asumir que los escenarios tienen una probabilidad igual de aparecer.Incluyendo los tres casos se puede hacer uno general, este diseño de programación lineal se le conoce como la forma extensiva del diseño estocástico debido a que describe explícitamente las variables de decisión de todos los escenarios.

Este caso se puede modelar de la siguiente manera:

$minimizar$

$150 x_1 + 230 x_2 + 260 x_3$

$+ \frac{1}{3}   ( 238 y_{11} + 210 y_{21} - 170 w_{11} - 150 w_{21}- 36 w_{31} - 10 w_{41})$ 

$+ \frac{1}{3}   ( 238 y_{12} + 210 y_{22} - 170 w_{12} - 150 w_{22}- 36 w_{32} - 10 w_{42})$ 

$+ \frac{1}{3}   ( 238 y_{13} + 210 y_{23} - 170 w_{13} - 150 w_{23}- 36 w_{33} - 10 w_{43})$

$tal$ $que$

$x_1 + x_2 + x_3 \leq 500$



$3 x_1 + y_{11} - w_{11} \geq 200$

$3.6 x_2 + y_{21} - w_{21} \geq 240$

$w_{31} + w_{41} \leq 24 x_3$

$w_{31} \leq 6000$


$2.5 x_1 + y_{12} - w_{12} \geq 200$

$3 x_2 + y_{22} - w_{22} \geq 240$

$w_{32} + w_{42} \leq 20 x_3$

$w_{32} \leq 6000$


$2 x_1 + y_{13} - w_{13} \geq 200$

$2.4 x_2 + y_{23} - w_{23} \geq 240$

$w_{33} + w_{43} \leq 16 x_3$

$w_{33} \leq 6000$


$x, y, w \geq 0$

In [4]:
#Tres escenarios simultáneos
#Programa: Problema del agricultor (granjero)


#Usar librerias
from pulp import *

#Se crea la variable de problema:
prob=LpProblem("Problema del agricultor (granjero)", LpMinimize)

#Se crean cada una de las variables que se requieren determinar
#Aquí se definen los límites de las variables:
x1=LpVariable("x1",lowBound=0)
x2=LpVariable("x2",lowBound=0)
x3=LpVariable("x3",lowBound=0)

y1_Bueno=LpVariable("y1 Bueno",lowBound=0)
y2_Bueno=LpVariable("y2 Bueno",lowBound=0)

y1_Prom=LpVariable("y1 Prom",lowBound=0)
y2_Prom=LpVariable("y2 Prom",lowBound=0)

y1_Malo=LpVariable("y1 Malo",lowBound=0)
y2_Malo=LpVariable("y2 Malo",lowBound=0)

w1_Bueno=LpVariable("w1 Bueno",lowBound=0)
w2_Bueno=LpVariable("w2 Bueno",lowBound=0)
w3_Bueno=LpVariable("w3 Bueno",lowBound=0)
w4_Bueno=LpVariable("w4 Bueno",lowBound=0)

w1_Prom=LpVariable("w1 Prom",lowBound=0)
w2_Prom=LpVariable("w2 Prom",lowBound=0)
w3_Prom=LpVariable("w3 Prom",lowBound=0)
w4_Prom=LpVariable("w4 Prom",lowBound=0)

w1_Malo=LpVariable("w1 Malo",lowBound=0)
w2_Malo=LpVariable("w2 Malo",lowBound=0)
w3_Malo=LpVariable("w3 Malo",lowBound=0)
w4_Malo=LpVariable("w4 Malo",lowBound=0)

#Se crea la funcion objetivo:
funcion_objetivo= 150*x1 + 230*x2 + 260*x3 + 1/3*(238*y1_Bueno + 210*y2_Bueno - 170*w1_Bueno - 150*w2_Bueno- 36*w3_Bueno - 10*w4_Bueno) + 1/3*(238*y1_Prom + 210*y2_Prom - 170*w1_Prom - 150*w2_Prom- 36*w3_Prom - 10*w4_Prom) + 1/3*(238*y1_Malo + 210*y2_Malo - 170*w1_Malo - 150*w2_Malo- 36*w3_Malo - 10*w4_Malo)

#La función objetivo se agrega al problema
prob += funcion_objetivo

#Se crean las restricciones (notese que se aumenta el número de restricciones debido a que hay más limitantes
restriccion1= x1 + x2 + x3 <= 500

restriccion2= 3*x1 + y1_Bueno - w1_Bueno >= 200
restriccion3= 3.6*x2 + y2_Bueno - w2_Bueno >= 240
restriccion4= w3_Bueno + w4_Bueno - 24*x3 <= 0
restriccion5= w3_Bueno <= 6000

restriccion6= 2.5*x1 + y1_Prom - w1_Prom >= 200
restriccion7= 3*x2 + y2_Prom - w2_Prom >= 240
restriccion8= w3_Prom + w4_Prom - 20*x3 <= 0
restriccion9= w3_Prom <= 6000

restriccion10= 2*x1 + y1_Malo - w1_Malo >= 200
restriccion11= 2.4*x2 + y2_Malo - w2_Malo >= 240
restriccion12= w3_Malo + w4_Malo - 16*x3 <= 0
restriccion13= w3_Malo <= 6000



#Las restricciones se añaden
prob += restriccion1
prob += restriccion2
prob += restriccion3
prob += restriccion4
prob += restriccion5
prob += restriccion6
prob += restriccion7
prob += restriccion8
prob += restriccion9
prob += restriccion10
prob += restriccion11
prob += restriccion12
prob += restriccion13

#Se soluciona el problema:
prob.solve()

#Resultados impresos:
print ("Status:", LpStatus[prob.status])
for v in prob.variables():
    print(v.name, "=", v.varValue)
print("Valor Óptimo: $", value(prob.objective))

Status: Optimal
w1_Bueno = 310.0
w1_Malo = 140.0
w1_Prom = 225.0
w2_Bueno = 48.0
w2_Malo = 0.0
w2_Prom = 0.0
w3_Bueno = 6000.0
w3_Malo = 4000.0
w3_Prom = 5000.0
w4_Bueno = 0.0
w4_Malo = 0.0
w4_Prom = 0.0
x1 = 170.0
x2 = 80.0
x3 = 250.0
y1_Bueno = 0.0
y1_Malo = 0.0
y1_Prom = 0.0
y2_Bueno = 0.0
y2_Malo = 48.0
y2_Prom = 0.0
Valor Óptimo: $ -108390.0


Si se supiera de antemano el resultado de los tres casos anteriores se puede usar un valor esperado para observar la cantidad que se espera  obtener, esta estaría dada por la suma de cada valor obtenido multiplicado por la probabilidad, debido  a que la probabilidad es uniforme se puede expresar como:

$ \frac{1}{3}( \$ 59950.0+ \$ 167666.66709+ \$ 118600.00) =  \$ 115405.5557$ 

En contraste con el resultado obtenido del valor optimo  considerando los tres casos, se puede obtener el siguiente valor: 

$ \$ 115405.5557- \$ 108390.0= \$7015.555679$. Este valor se le conoce como el *valor esperado de la información perfecta.*
